### ISSUE TITLE
accelerate-cuda-0.15.1.1 ghci "out of memory error"

### ISSUE THREAD

**Issue by cpdurham** [AccelerateHS/accelerate#297]
Got accelerate-cuda-0.15.1.1 working no problem compiled (awesome!)

However, had a "**\* Exception: CUDA Exception: out of memory" error when run from ghci

did 

```
stack ghci accelerate-cuda-0.15.1.1 cuda-0.7.0.0 language-c-quote-0.11.4 mainland-pretty-0.4.1.2 exception-mtl-0.4 symbol-0.2.4
```

and then did 

```
Prelude> import Data.Array.Accelerate as A
Prelude A> import Data.Array.Accelerate.CUDA as C
Prelude A C> C.run1 (A.map (+1)) $ A.fromList (Z :. 10 :. 10 :: DIM2) ([0..] :: [Int])
```

and got

```
*** Exception: CUDA Exception: out of memory
```

Same thing (same deps) worked just fine compiled

**Comment by tmcdonell** [comment]
This is strange. The only thing I can think of is that for some reason there didn't happen to be enough _contiguous_ memory available for the allocation. But it is such a tiny array, and if you consistently get this problem, so this seem unlikely...

I will see if I can reproduce in 0.15 and get back to you...

**Comment by tmcdonell** [comment]
Oh, was this with GHC-8 by chance?

**Comment by cpdurham** [comment]
No, I'm pretty sure it was ghc 7.10, I've never used 8 yet, so it couldn't have been 8.

**Comment by radoye** [comment]
Potentially relevant info. 
With GHC7.10 everything seems to work. With GHC8 "out of memory" is reported right away (as GHC is starting up and `props 0` returns either "out of memory" or the error at the bottom. 

```
rrs@elsuperbeasto> stack repl                                                                 ~/workspace/dev/haskell/lab/accl
Downloaded lts-6.26 build plan.    
[1 of 2] Compiling Main             ( /home/rrs/.stack/setup-exe-src/setup-mPHDZzAJ.hs, /home/rrs/.stack/setup-exe-src/setup-mPHDZzAJ.o )
[2 of 2] Compiling StackSetupShim   ( /home/rrs/.stack/setup-exe-src/setup-shim-mPHDZzAJ.hs, /home/rrs/.stack/setup-exe-src/setup-shim-mPHDZzAJ.o )
Linking /home/rrs/.stack/setup-exe-cache/x86_64-linux-ncurses6/tmp-Cabal-simple_mPHDZzAJ_1.22.5.0_ghc-7.10.3 ...
[...]
cuda-0.7.5.1: configure
accelerate-0.15.1.0: copy/register
cuda-0.7.5.1: build
cuda-0.7.5.1: copy/register
accl-0.1.0.0: configure (lib)
Configuring accl-0.1.0.0...
accl-0.1.0.0: initial-build-steps (lib)
Completed 20 action(s).
Warning: File listed in accl.cabal file does not exist: README.md
Configuring GHCi with the following packages: accl
GHCi, version 7.10.3: http://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling Lib              ( /home/rrs/workspace/dev/haskell/lab/accl/src/Lib.hs, interpreted )
Ok, modules loaded: Lib.
*Lib Lib> :m +Foreign.CUDA
*Lib Lib Foreign.CUDA> props 0
DeviceProperties {deviceName = "GeForce GTX 650", computeCapability = 3.0, totalGlobalMem = 1055260672, totalConstMem = 65536, sharedMemPerBlock = 49152, regsPerBlock = 65536, warpSize = 32, maxThreadsPerBlock = 1024, maxThreadsPerMultiProcessor = 2048, maxBlockSize = (1024,1024,64), maxGridSize = (2147483647,65535,65535), maxTextureDim1D = 65536, maxTextureDim2D = (65536,65536), maxTextureDim3D = (4096,4096,4096), clockRate = 1058500, multiProcessorCount = 2, memPitch = 2147483647, memBusWidth = 128, memClockRate = 2500000, textureAlignment = 512, computeMode = Default, deviceOverlap = True, concurrentKernels = True, eccEnabled = False, asyncEngineCount = 1, cacheMemL2 = 262144, pciInfo = PCI {busID = 1, deviceID = 0, domainID = 0}, tccDriverEnabled = False, kernelExecTimeoutEnabled = True, integrated = False, canMapHostMemory = True, unifiedAddressing = True, streamPriorities = False, globalL1Cache = False, localL1Cache = True, managedMemory = True, multiGPUBoard = False, multiGPUBoardGroupID = 0}
*Lib Lib Foreign.CUDA> 
Leaving GHCi.
rrs@elsuperbeasto> nvidia-device-query                                                        ~/workspace/dev/haskell/lab/accl
CUDA device query (Driver API, statically linked)
CUDA driver version 8.0
Detected 1 CUDA capable device

Device 0: GeForce GTX 650
  CUDA capability:                    3.0
  CUDA cores:                         384 cores in 2 multiprocessors (192 cores/MP)
  Global memory:                      1006 MB
  Constant memory:                    64 kB
  Shared memory per block:            48 kB
  Registers per block:                65536
  Warp size:                          32
  Maximum threads per multiprocessor: 2048
  Maximum threads per block:          1024
  Maximum grid dimensions:            2147483647 x 65535 x 65535
  Maximum block dimensions:           1024 x 1024 x 64
  GPU clock rate:                     1.0585 GHz
  Memory clock rate:                  2.5 GHz
  Memory bus width:                   128-bit
  L2 cache size:                      256 kB
  Maximum texture dimensions         
    1D:                               65536
    2D:                               65536 x 65536
    3D:                               4096 x 4096 x 4096
  Texture alignment:                  512 B
  Maximum memory pitch:               2 GB
  Concurrent kernel execution:        Yes
  Concurrent copy and execution:      Yes, with 1 copy engine
  Runtime limit on kernel execution:  Yes
  Integrated GPU sharing host memory: No
  Host page-locked memory mapping:    Yes
  ECC memory support:                 No
  Unified addressing (UVA):           Yes
  PCI bus/location:                   1/0
  Compute mode:                       Default
    Multiple contexts are allowed on the device simultaneously
rrs@elsuperbeasto>                                                                            ~/workspace/dev/haskell/lab/accl
```

Changing the resolver from `lts-6.26` to `lts-7.8` results in
```
rrs@elsuperbeasto> vim stack.yaml                                                                                                              
rrs@elsuperbeasto> stack repl                                                                 ~/workspace/dev/haskell/lab/accl
accl-0.1.0.0: configure (lib)
Configuring accl-0.1.0.0...
accl-0.1.0.0: initial-build-steps (lib)
Configuring GHCi with the following packages: accl
GHCi, version 8.0.1: http://www.haskell.org/ghc/  :? for help
Failed to pre-initialise CUDA: out of memory
[1 of 1] Compiling Lib              ( /home/rrs/workspace/dev/haskell/lab/accl/src/Lib.hs, interpreted )
Ok, modules loaded: Lib.
Loaded GHCi configuration from /tmp/ghci19808/ghci-script
*Lib Lib> :m +Foreign.CUDA
*Lib Lib Foreign.CUDA> props 0
*** Exception: ComputeMode.toEnum: Cannot match 32713
CallStack (from HasCallStack):
  error, called at src/Foreign/CUDA/Analysis/Device.chs:63:22 in cuda-0.7.5.1-JchsevDeL2OEazVWFBfQs:Foreign.CUDA.Analysis.Device
*Lib Lib Foreign.CUDA> 
Leaving GHCi.
rrs@elsuperbeasto>     
```

**Comment by tmcdonell** [comment]
GHC-8.0.1 failure is definitely due to [GHC#12573](https://ghc.haskell.org/trac/ghc/ticket/12573), which will be fixed in 8.0.2. Compiled programs should still work with 8.0.1 though.

**Comment by tmcdonell** [comment]
The relevant GHC bug is fixed in 8.0.2 and onwards; closing.
